debian/sail-convert.1
'\" t
.\"     Title: sail-convert
.\"    Author: Dmitry Baryshev
.\"      Date: 05/14/2020
.\"    Manual: SAIL Manual
.\"    Source: SAIL
.\"  Language: English
.\"
.TH "SAIL\-CONVERT" "1" "05/14/2020" "SAIL" "SAIL Manual"
.SH "NAME"
sail-convert \- SAIL-based tool to convert between image formats
.SH "SYNOPSIS"
.sp
\fBsail\-convert\fR \fI<input file>\fR \fI<output file>\fR [options] | -h|--help | -v|--version
.SH "DESCRIPTION"
.sp
\fBsail\-convert\fR is used to convert images from one format to another with a simple set of options.
.SH "OPTIONS"
.PP
\-c | --compression <value>
.RS 4
Use the specified compression value to write the output image. For example, for JPEGs the value could be in range [0..100].
.RE
.PP
\-v | --version
.RS 4
Print the version and exit.
.RE
.PP
\-h | --help
.RS 4
Print the help summary and exit.
.RE
.SH "EXIT STATUS"
.sp
\fBsail\-convert\fR returns 0 on success and a SAIL error code otherwise.
.SH "COPYRIGHT NOTICE"
Copyright \(co 2020 Dmitry Baryshev



CHANGELOG.md
# ChangeLog

SAIL (Squirrel Abstract Image Libraries) is a fork of ksquirrel-libs which was a set of C++ image codecs for KSquirrel image viewer.
See [http://ksquirrel.sourceforge.net](http://ksquirrel.sourceforge.net).

See [README](README.md) for more.

# SAIL (since 0.9.0)

## 0.9.0 XXXX-XX-XX

- first public release
- implemented rich C client API to read and write images
- implemented rich C++ client API to read and write images
- plugins interfaces are now hidden. Always use the client APIs to read or write images

# ksquirrel-libs (until 0.8.0)

## 0.8.0 2007-12-08

- added EPS (reading)
- added PSP (reading)
- 59 codecs in total
- bugfix in PNM decoder on P3 PNMs
- bugfix in CAMERA codec, when KSquirrel couldn't open raw fotos

## 0.7.5 2007-11-14

- CAMERA codec now uses dcraw v8.79
- update emails to ksquirrel.iv@gmail.com

## 0.7.4 2007-11-05

- no changes

## 0.7.3 2007-11-04

- added DDS (reading)
- 57 codecs in total
- JPEG2000 decoder is much faster now
- check for errors when external converter programs are executed

## 0.7.2 2007-10-17

- added DICOM (reading, requires medcon)
- added APNG (reading, see [http://en.wikipedia.org/wiki/APNG](http://en.wikipedia.org/wiki/APNG))
- 56 codecs in total
- fixed rare crashes with some GIFs
- added LJPEG missing data
- added TIFF settings UI
- SVG codec now uses rsvg-convert
- code cleanup

## 0.7.1-try3 2007-09-21

- fixed random crashes with GIFs

## 0.7.1-try2 2007-09-14

- added LJPEG (reading)
- 54 codecs in total
- configure script now checks for libungif

## 0.7.1 2007-08-30

- added XFIG (reading, transfig package is required)
- added Degas PI3 (reading)
- 53 codecs in total
- added mime types
- TTF codec is designed for freetype 2.2.1 (it should work fine at least in Debian Etch)
- CAMERA codec now uses dcraw v8.77 without modifications, so it doesn't violate dcraw license anymore
- bugfixes in scripts

## 0.7.0 2007-05-13

- updated Qt examples

## 0.7.0-preview2 2007-04-09

- added LEAF, NEO, PI1, PICT, XIM via NetPBM package
- 51 codecs in total
- small fixes in configure.ac

## 0.7.0-preview1 2007-03-09

- added DjVu (reading, just single page via settings)
- added AutoCAD DXF (reading, +settings)
- added GIMP XCF (reading, +settings)
- 46 codecs in total
- fixed random segfaults on GIFs and some other codecs
- CAMERA codec now uses dcraw 8.61
- added settings support to CAMERA, SVG
- small fixes in configure script
- configure output is more verobose now

## 0.6.2 2006-01-02

- API changes

## 0.6.1 2005-12-21

- MAC, MSP, IFF, and UTAH codecs are disabled as buggy (use --enable-devel to enable them)
- ksquirrel-libs now installs a development library and header files
- ksquirrel-libs now must be installed before KSquirrel
- updated documentation

## 0.6.0 2005-12-10

- SVG: added ".svgz" extension
- MNG: small fix in data types
- KOALA: small fix in bitdepth (bitdepth was always '0')
- security fixes

## 0.6.0-preview9 2005-11-09

- added TTF fonts support (ttf, pfa, pfb...)
- added MNG, JNG (reading)
- added JBIG (slow, reading))
- 43 codecs in total
- XCUR: fixed segfault
- PNM: small fix on pgm images
- SGI: small improvements
- CAMERA: new file extensions added
- code cleanup
- no more static libraries. Now SVG, GIF, OPENEXR, WMF, and JPEG2000 are optional

## 0.6.0-preview8 2005-10-10

- autoconf script cleanup and bugfixes
- new options in configure script:
    `--disable-gif`
    `--disable-camera`
    `--disable-wmf`
    `--disable-svg`
    `--disable-openexr`
    `--disable-jpeg2000`
    `--disable-mng`
    `--disable-ttf`

## 0.6.0-preview7 2005-10-02

- added PXR (reading)
- 40 codecs in total
- code cleanup
- added numerous image filtering functions

## 0.6.0-preview6 2005-08-31

- added MTV Ray-Tracer (reading, writing)
- added AVS X (reading, writing)
- added MAC (reading)
- added MSP (reading)
- 39 codecs in total
- added writing features to PNM
- some changes in the API
- LIF: many bugfixes
- SVG: removed useless "-lfontconfig" dependency which could lead to compile errors on some systems (e.g. Slackware)
- PNG: fixed autoconf errors on some systems (e.g. Slackware)
- added new image filtering functions

## 0.6.0-preview5 2005-08-11

- added JPEG2000 (reading)
- added OpenEXR (reading)
- added WAL Quake2 texture (reading)
- added MDL HalfLife model (reading)
- added KOALA (reading)
- added HDR (reading)
- added LIF (reading)
- added SCT (reading)
- 35 codecs in total
- CUT decoder is now not-alpha

## 0.6.0-preview4 2005-06-26

- cosmetic changes

## 0.6.0-preview3 2005-05-27

- added SVG (reading, requires libxml2, freetype)
- 27 codecs in total
- some changes in the API makes prevew2 and prevew3 not compatible
- TGA: small fix (in v0.7.1 flipping was ON by default, in v0.7.2 flipping is determined by image header)
- XPM: small fix in decoding mechanism
- new mime icons
- added image filtering C++ utils

## 0.6.0-preview2 2005-04-17

- added CUT (reading)
- added IFF (reading)
- added WMF (reading)
- added SUN Icon (reading)
- added WBMP (reading)
- added TIFF (writing)
- added RAWRGB (reading, writing)
- added RAW (photos from different cameras, e.g. CRW etc.) through dcraw
- 26 codecs in total
- changed installation path /usr/lib/squirrel -> /usr/lib/ksquirrel-libs
- fully migrated to C++ and classes
- added some examples (Qt, Qt+OpenGL)

## 0.6.0-preview1 2005-02-28

- added interface for write functions
- added write features for PNG, JPEG, BMP
- PNG: fixed issues with interlaced images, fixed memory leaks
- XPM: fixed issues with multiline comments
- ICO: added support of bit depth 24 and 32
- PSD: fixed issues with RGB images which have 3 channels instead of 4
- GIF: added comments support (comment extensions)
- PNM: fixed issues with Windows-like line breaks (\r\n)

## 0.5.0 2005-02-20

- added X cursors
- added Adobe PSD (RGB,CMYK,Grayscale,Indexed)
- added FLI Animation
- 18 codecs in total
- GIF decoder is much better now (fixed issues with transparency)
- small fixes in PNM and BMP codecs

examples/c/sail-convert/sail-convert.c
/*
    Copyright (c) 2020 Dmitry Baryshev

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
*/

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "sail-common.h"
#include "sail.h"

static sail_error_t convert(const char *input, const char *output, struct sail_context *context, int compression) {

    SAIL_CHECK_PATH_PTR(input);
    SAIL_CHECK_PATH_PTR(output);
    SAIL_CHECK_CONTEXT_PTR(context);

    const struct sail_plugin_info *plugin_info;
    void *state;

    struct sail_image *image;
    void *image_bits;

    /* Read the image. */
    SAIL_LOG_INFO("Input file: %s", input);

    SAIL_TRY(sail_plugin_info_from_path(input, context, &plugin_info));
    SAIL_LOG_INFO("Input codec: %s", plugin_info->description);

    SAIL_TRY(sail_start_reading_file(input, context, plugin_info, &state));

    SAIL_TRY(sail_read_next_frame(state, &image, &image_bits));
    SAIL_TRY(sail_stop_reading(state));

    /* Write the image. */
    SAIL_LOG_INFO("Output file: %s", output);

    SAIL_TRY(sail_plugin_info_from_path(output, context, &plugin_info));
    SAIL_LOG_INFO("Output codec: %s", plugin_info->description);

    struct sail_write_options *write_options;
    SAIL_TRY(sail_alloc_write_options_from_features(plugin_info->write_features, &write_options));

    /* Apply our tuning. */
    SAIL_LOG_INFO("Compression: %d%s", compression, compression == -1 ? " (default)" : "");
    write_options->compression = compression;

    SAIL_TRY(sail_start_writing_file_with_options(output, context, plugin_info, write_options, &state));
    SAIL_TRY(sail_write_next_frame(state, image, image_bits));
    SAIL_TRY(sail_stop_writing(state));

    /* Clean up. */
    free(image_bits);
    sail_destroy_write_options(write_options);
    sail_destroy_image(image);

    SAIL_LOG_INFO("Success");

    return 0;
}

static void help(char *app) {

    fprintf(stderr, "sail-convert: Convert one image format to another.\n\n");
    fprintf(stderr, "Usage: %s <PATH TO INPUT IMAGE> <PATH TO OUTPUT IMAGE> [-c | --compression <value>]\n", app);
    fprintf(stderr, "       %s [-v | --version]\n", app);
    fprintf(stderr, "       %s [-h | --help]\n", app);
}

int main(int argc, char *argv[]) {

    if (argc < 2) {
        help(argv[0]);
        return 1;
    }

    if (strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0) {
        help(argv[0]);
        return 0;
    }

    if (strcmp(argv[1], "-v") == 0 || strcmp(argv[1], "--version") == 0) {
        fprintf(stderr, "sail-convert 1.0\n");
        return 0;
    }

    if (argc < 3) {
        help(argv[0]);
        return 1;
    }

    /* -1: default compression will be selected. */
    int compression = -1;

    /* Start parsing CLI options from the third argument. */
    int i = 3;

    while (i < argc) {
        if (strcmp(argv[i], "-c") == 0 || strcmp(argv[i], "--compression") == 0) {
            if (i == argc-1) {
                fprintf(stderr, "Error: Missing compression value.\n");
                return 1;
            }

            compression = atoi(argv[i+1]);
            i += 2;
            continue;
        }

        fprintf(stderr, "Error: Unrecognized option '%s'.\n", argv[i]);
        return 1;
    }

    struct sail_context *context;

    SAIL_TRY(sail_init(&context));

    SAIL_TRY(convert(argv[1], argv[2], context, compression));

    sail_finish(context);

    return 0;
}
